resource/src/test/java/io/smallrye/common/resource/JarFileResourceLoaderTests.java

```diff
@@ -0,0 +1,154 @@
+package io.smallrye.common.resource;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.DirectoryStream;
+import java.util.Iterator;
+import java.util.zip.CRC32;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import org.junit.jupiter.api.Test;
+
+public class JarFileResourceLoaderTests {
+
+    public static final String FILE_TXT = "This is a plain text file\nIf you can read this, it's working!\n";
+
+    @Test
+    public void testOpenJar() throws IOException {
+        try (JarFileResourceLoader rl = makeJar(
+                // keep these two first
+                dir("META-INF"),
+                deflate("META-INF/MANIFEST.MF", "Manifest-Version: 1.0\n"),
+                // test entries
+                dir("dir1"),
+                store("dir1/file-stored.txt", FILE_TXT),
+                deflate("dir1/file-deflated.txt", FILE_TXT),
+                // keep this as last entry
+                dir("end"))) {
+            assertNull(rl.findResource("missing"));
+            Resource dir1_file_stored_txt = rl.findResource("dir1/file-stored.txt");
+            assertNotNull(dir1_file_stored_txt);
+            assertEquals(FILE_TXT, dir1_file_stored_txt.asString(StandardCharsets.UTF_8));
+            Resource dir1_file_deflated_txt = rl.findResource("dir1/file-deflated.txt");
+            assertNotNull(dir1_file_deflated_txt);
+            assertEquals(FILE_TXT, dir1_file_deflated_txt.asString(StandardCharsets.UTF_8));
+            Resource dir1 = rl.findResource("dir1");
+            assertNotNull(dir1);
+            assertEquals(0, dir1.size());
+            try (DirectoryStream<Resource> ds = dir1.openDirectoryStream()) {
+                Iterator<Resource> iterator = ds.iterator();
+                assertTrue(iterator.hasNext());
+                assertEquals("dir1/file-stored.txt", iterator.next().pathName());
+                assertTrue(iterator.hasNext());
+                assertEquals("dir1/file-deflated.txt", iterator.next().pathName());
+                assertFalse(iterator.hasNext());
+            }
+        }
+    }
+
+    private static JarFileResourceLoader makeJar(Entry... entries) throws IOException {
```

Comment:
We have `shrinkwrap` as a test dependency in other modules to help with this if you prefer. We can also use it here.

Reply:
I prefer to keep it simple in this case, since these tests may require detailed control over the archive format for specific reasons.